### PR TITLE
Import kotlinx.serialization converter

### PR DESCRIPTION
**Pull request by JakeWharton**
From https://github.com/JakeWharton/retrofit2-kotlinx-serialization-converter/. Now that all the APIs and types in use are stable, we can ship this as a first-party converter.

### CODE REVIEW

File: retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Factory.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    return SerializationStrategyConverter(contentType, saver, serializer)
+  }
+}
+
+/**
+ * Return a [Converter.Factory] which uses Kotlin serialization for string-based payloads.
+ *
+ * Because Kotlin serialization is so flexible in the types it supports, this converter assumes
+ * that it can handle all types. If you are mixing this with something else, you must add this
+ * instance last to allow the other converters a chance to see their types.
```
Comment (swankjesse) [inline retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Factory.kt:47]: 👍🏻

File: retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Factory.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+/**
+ * Return a [Converter.Factory] which uses Kotlin serialization for byte-based payloads.
+ *
+ * Because Kotlin serialization is so flexible in the types it supports, this converter assumes
+ * that it can handle all types. If you are mixing this with something else, you must add this
+ * instance last to allow the other converters a chance to see their types.
+ */
+@JvmName("create")
```
Comment (swankjesse) [inline retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Factory.kt:61]: I wonder how big the intersection is between users of Kotlin serialization and users of Java. Probably worth including anyway for projects migrating.

File: retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Serializer.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    override fun <T> toRequestBody(contentType: MediaType, saver: SerializationStrategy<T>, value: T): RequestBody {
+      val string = format.encodeToString(saver, value)
+      return RequestBody.create(contentType, string)
+    }
+  }
+
+  class FromBytes(override val format: BinaryFormat) : Serializer() {
+    override fun <T> fromResponseBody(loader: DeserializationStrategy<T>, body: ResponseBody): T {
+      val bytes = body.bytes()
```
Comment (swankjesse) [inline retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Serializer.kt:37]: In theory you could also hook up the kotlinx.serialization Okio integration to get streaming. This is fine though.

File: retrofit-converters/kotlinx-serialization/src/test/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/KotlinSerializationConverterFactoryBytesTest.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import retrofit2.Call
+import retrofit2.Retrofit
+import retrofit2.http.Body
+import retrofit2.http.GET
+import retrofit2.http.POST
+
+private val bobBytes = ByteString.of(0x0a, 0x03, 'B'.code.toByte(), 'o'.code.toByte(), 'b'.code.toByte())
```
Comment (swankjesse) [inline retrofit-converters/kotlinx-serialization/src/test/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/KotlinSerializationConverterFactoryBytesTest.kt:22]: 😭

File: retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Serializer.kt (hunk 3 above)
Comment (JakeWharton) [reply retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Serializer.kt:37]: Streaming is only available for JSON. Not arbitrary `BinaryFormat` or `TextFormat`s.

File: retrofit-converters/kotlinx-serialization/src/test/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/KotlinxSerializationConverterFactoryContextualTest.kt [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    @GET("/")
+    fun deserialize(): Call<User>
+
+    @POST("/")
+    fun serialize(@Body user: User): Call<Void?>
+  }
+
+  data class User(val name: String)
+
+  object UserSerializer : KSerializer<User> {
```
Comment (swankjesse) [inline retrofit-converters/kotlinx-serialization/src/test/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/KotlinxSerializationConverterFactoryContextualTest.kt:41]: Why hand-write this?

Comment (swankjesse) [reply retrofit-converters/kotlinx-serialization/src/test/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/KotlinxSerializationConverterFactoryContextualTest.kt:41]: Just to confirm it all gets wired in? That’s okay.

**Review by swankjesse — APPROVED** [review]

File: retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Serializer.kt (hunk 3 above)
Comment (swankjesse) [reply retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Serializer.kt:37]: Ooooh that’s a thing for them to fix.

I expect JSON is like 90% of kotlinx.serialization usage. It’s great that it supports XML and protobuf and others, but I expect JSON is where specializations like this are most likely to pay off.

Definitely not for this PR.

File: retrofit-converters/kotlinx-serialization/src/test/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/KotlinxSerializationConverterFactoryContextualTest.kt (hunk 5 above)
Comment (JakeWharton) [reply retrofit-converters/kotlinx-serialization/src/test/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/KotlinxSerializationConverterFactoryContextualTest.kt:41]: Yeah I believe before we were looking up the serializer in a way that did not honor contextually-registered ones.

**Comment by Goooler** [comment]
Should we align packages `com.jakewharton.retrofit2` to `retrofit2`?

**Comment by JakeWharton** [comment]
Yes, whoops

**Pull request merged**